### PR TITLE
Add Support for `java:identifier` Metadata

### PR DESCRIPTION
**Pull request by InsertCreativityHere**
This PR adds support for `java:identifier`, just as I've already done for C++ and C#.
It cannot be applied to modules for the same reason we haven't added it in C++ and C# yet.

But, a difference in Java is that `java:identifier` cannot be applied to classes or exceptions do to how the Java activator works.
It expects a tight coupling between a class's typeID and Java class name. `java:identifier` completely decouples these.

And since I've removed the auto-escaping logic, any users with classes or exceptions that collide with Java keywords will be unable to upgrade past 3.7. I expect this is a very rare occurrence though. See #3534 for further discussion.

### CODE REVIEW

File: cpp/src/slice2java/Gen.h [hunk 1]
```diff
@@ -49,7 +49,6 @@ namespace Slice
         //
         // Compose the argument lists for an operation.
         //
-        std::vector<std::string> getArgs(const OperationPtr&);
```
Comment (InsertCreativityHere) [inline cpp/src/slice2java/Gen.h:52]: Dead function. Never called.

File: cpp/src/slice2java/Gen.h [hunk 2]
```diff
@@ -117,7 +116,7 @@ namespace Slice
     class Gen final
     {
     public:
-        Gen(const std::string&, std::string, const std::vector<std::string>&, std::string);
```
Comment (InsertCreativityHere) [inline cpp/src/slice2java/Gen.h:120]: For some reason we were passing in the executable name (`slice2java`) here.
We never checked or used this field though. Dead.

File: cpp/src/slice2java/JavaUtil.cpp [hunk 3]
```diff
@@ -343,83 +319,6 @@ Slice::JavaGenerator::output() const
...
-    stringstream result;
-    for (const auto& id : ids)
-    {
-        result << "::" + id;
-    }
-    return result.str();
-}
-
-string
-Slice::JavaGenerator::convertScopedName(const string& scoped, const string& prefix, const string& suffix)
```
Comment (InsertCreativityHere) [inline cpp/src/slice2java/JavaUtil.cpp:375]: Now that `mappedScoped` has a 'separator' argument, we can let `libSlice` give us the separator we want (`.`),
instead of needing this function to massage it after-the-fact.

File: cpp/src/slice2java/JavaUtil.h [hunk 4]
```diff
@@ -114,11 +104,7 @@ namespace Slice
         // package argument matches the entity's package name, then the
         // package is removed from the result.
         //
-        [[nodiscard]] static std::string getUnqualified(
-            const ContainedPtr&,
-            const std::string& = std::string(),
-            const std::string& = std::string(),
-            const std::string& = std::string());
```
Comment (InsertCreativityHere) [inline cpp/src/slice2java/JavaUtil.h:121]: We no longer pass `prefix` and `suffix` in here. Instead we just add them where we need them.

File: java/src/com.zeroc.ice/src/main/java/com/zeroc/Ice/Util.java [hunk 5]
```diff
@@ -528,7 +528,7 @@ public static String typeIdToClass(String id) {
             if (buf.length() > 0) {
                 buf.append('.');
             }
-            buf.append(fixKwd(s));
```
Comment (InsertCreativityHere) [inline java/src/com.zeroc.ice/src/main/java/com/zeroc/Ice/Util.java:531]: A result of this change is that if a user has a class/exception that collides with a Java keyword,
they will be stuck using 3.7. This is not optimal.

File: java/test/src/main/java/test/Ice/adapterDeactivation/Test.ice [hunk 6]
```diff
@@ -8,6 +8,7 @@ module Test
 
 interface TestIntf
 {
+    ["java:identifier:_transient"]
     void transient();
```
Comment (InsertCreativityHere) [inline java/test/src/main/java/test/Ice/adapterDeactivation/Test.ice:12]: I kept the operation name the same because this is what we call it in every other language's `adapterDeactivation` test.

File: java/test/src/main/java/test/Ice/packagemd/NoPackage.ice [hunk 7]
```diff
@@ -25,7 +25,7 @@ exception E2 extends E1
     long l;
 }
 
-exception notify /* Test keyword escape. */
```
Comment (InsertCreativityHere) [inline java/test/src/main/java/test/Ice/packagemd/NoPackage.ice:28]: `notify` is not a keyword, but it is a 'special' identifier, defined on every Object.
Here, there _should_ be no problem with it as an exception name though.

File: cpp/src/slice2java/Main.cpp [hunk 8]
```diff
@@ -264,7 +264,7 @@ compile(const vector<string>& argv)
                 {
                     try
                     {
-                        Gen gen(argv[0], icecpp->getBaseName(), includePaths, output);
```
Comment (InsertCreativityHere) [inline cpp/src/slice2java/Main.cpp:267]: The executable name we were never using...

File: cpp/src/slice2java/Gen.cpp [hunk 9]
```diff
@@ -1474,22 +1446,17 @@ Slice::JavaVisitor::writeDispatch(Output& out, const InterfaceDefPtr& p)
...
         {
             out << nl << "request.inputStream.skipEmptyEncapsulation();";
         }
 
-        vector<string> inArgs;
-        for (const auto& pli : inParams)
-        {
-            inArgs.push_back("iceP_" + pli->name());
-        }
-
```
Comment (InsertCreativityHere) [inline cpp/src/slice2java/Gen.cpp:1492]: We already had a helper function for getting this list: `getInArgs`.

**Review by InsertCreativityHere — COMMENTED** [review]
Also, just for the record, `slice2java` only performs case-correction for niche things we generate. Specifically:
- <OP_NAME>MarshaledResult
- <OP_NAME>Result

- set<OPTIONAL_MEMBER_NAME>
- get<OPTIONAL_MEMBER_NAME>
- has<OPTIONAL_MEMBER_NAME>
- clear<OPTIONAL_MEMBER_NAME>
- optional<OPTIONAL_MEMBER_NAME>
- is<BOOL_MEMBER_NAME>

- set<INDEXABLE_MEMBER_NAME>
- get<INDEXABLE_MEMBER_NAME>

File: java/src/com.zeroc.ice/src/main/java/com/zeroc/Ice/Util.java (hunk 5 above)
Comment (bernardnormier) [reply java/src/com.zeroc.ice/src/main/java/com/zeroc/Ice/Util.java:531]: > if a user has a class/exception that collides with a Java keyword

Nobody has such a class or exception, so this is definitely not a concern.

File: cpp/src/slice2java/JavaUtil.h [hunk 10]
```diff
@@ -114,11 +104,7 @@ namespace Slice
...
+        [[nodiscard]] static std::string getUnqualified(const ContainedPtr&, const std::string& = std::string());
```
Comment (bernardnormier) [inline cpp/src/slice2java/JavaUtil.h:107]: Always good to add parameter names to signatures when you update the code.

File: cpp/src/slice2java/Gen.h (hunk 1 above)
Comment (bernardnormier) [reply cpp/src/slice2java/Gen.h:52]: I find the comments rather confusing. Compose the parameter lists vs compose the argument lists??

File: java/test/src/main/java/test/Ice/adapterDeactivation/Test.ice (hunk 6 above)
Comment (bernardnormier) [reply java/test/src/main/java/test/Ice/adapterDeactivation/Test.ice:12]: I don't like _transient: leading underscore is for private fields. I suggest `transient_`.

File: java/test/src/main/java/test/Ice/packagemd/NoPackage.ice (hunk 7 above)
Comment (bernardnormier) [reply java/test/src/main/java/test/Ice/packagemd/NoPackage.ice:28]: What are we testing here? If a test is redundant, it should be deleted.

File: java/test/src/main/java/test/Slice/escape/Key.ice [hunk 11]
```diff
@@ -4,71 +4,71 @@
 
 [["java:package:test.Slice.escape"]]
 
-module abstract
+// TODO: figure out a better way to map module names.
+module java_abstract
 {
 
+["java:identifier:_assert"]
```
Comment (bernardnormier) [inline java/test/src/main/java/test/Slice/escape/Key.ice:11]: I would pick somewhat recommendable remapping. identifiers with leading underscores or prefixed by java_ are not reasonable recommendations.

**Review by bernardnormier — APPROVED** [review]

File: cpp/src/slice2java/Gen.cpp [hunk 12]
```diff
@@ -1474,22 +1446,17 @@ Slice::JavaVisitor::writeDispatch(Output& out, const InterfaceDefPtr& p)
 
             for (const auto& value : values)
             {
-                const string typeS =
+                const string ts =
```
Comment (pepone) [inline cpp/src/slice2java/Gen.cpp:1449]: I would remove `ts` variable, and use typeToString directly. The variable is only used once and doesn't make the code any clear.

File: cpp/src/slice2java/Gen.cpp [hunk 13]
```diff
@@ -2148,11 +2100,13 @@ Slice::JavaVisitor::writeServantDocComment(Output& out, const OperationPtr& p, c
         out << nl << " * @return A completion stage that the servant will complete when the invocation completes.";
     }
 
-    map<string, StringList> exDocs = dc->exceptions();
-    for (const auto& exDoc : exDocs)
+    for (const auto& [name, lines] : dc->exceptions())
```
Comment (pepone) [inline cpp/src/slice2java/Gen.cpp:2103]: We have very similar code above, we should consider refactor to a function in a follow-up PR.

File: cpp/src/slice2java/Gen.cpp [hunk 14]
```diff
@@ -2343,13 +2296,13 @@ Slice::Gen::TypesVisitor::visitClassDefStart(const ClassDefPtr& p)
                 //
                 // Generate a constructor accepting parameters for just the required members.
                 //
-                out << sp << nl << "public " << fixKwd(name) << spar;
+                out << sp << nl << "public " << name << spar;
                 vector<string> parameters;
                 for (const auto& allDataMember : allDataMembers)
```
Comment (pepone) [inline cpp/src/slice2java/Gen.cpp:2301]: not related to the PR, but `allDataMember` seems an awful name for the iteration variable.

**Review by pepone — APPROVED** [review]

File: cpp/src/slice2java/Gen.h (hunk 1 above)
Comment (InsertCreativityHere) [reply cpp/src/slice2java/Gen.h:52]: While unrelated, I cleaned up much this code, and fixed the doc-comments accordingly.

File: cpp/src/slice2java/JavaUtil.h (hunk 10 above)
Comment (InsertCreativityHere) [reply cpp/src/slice2java/JavaUtil.h]: I added parameter names to all the functions in this area of the code, including this one.

File: java/test/src/main/java/test/Slice/escape/Key.ice (hunk 11 above)
Comment (InsertCreativityHere) [reply java/test/src/main/java/test/Slice/escape/Key.ice:11]: Typically I agree with this, but here we're talking about the `escape` test.
There isn't a single line of this file that is "reasonable" or "recommendable".

You could argue the whole point of this test is to make sure even the most insane edge-case names still work correctly.

Hopefully the `java_` prefixes are temporary, but we have no way to remap modules, classes, or exceptions right now.
And Slice forbids leading underscores. Sure, we could pick some totally unrelated name, but this would lose the intent.
These are _supposed_ to be tested, we just lack the ability to do so right now.

File: java/test/src/main/java/test/Ice/adapterDeactivation/Test.ice (hunk 6 above)
Comment (bernardnormier) [reply java/test/src/main/java/test/Ice/adapterDeactivation/Test.ice:12]: Just pick a different name then, such as `transientOp`.

File: java/test/src/main/java/test/Slice/escape/Key.ice (hunk 11 above)
Comment (bernardnormier) [reply java/test/src/main/java/test/Slice/escape/Key.ice:11]: I'd suggest to add an `escaped` for readability, as in `escapedAssert`.

It's weird to have java-prefixed identifiers in Java. Your language is Java, the identifiers are obviously Java identifiers.

File: java/test/src/main/java/test/Ice/packagemd/NoPackage.ice (hunk 7 above)
Comment (InsertCreativityHere) [reply java/test/src/main/java/test/Ice/packagemd/NoPackage.ice:28]: It's like I said, it seems like we're testing that you can throw and catch exceptions that have been re-mapped with `java:package`. If you're asking for the specific code snippet, here:
```
            try {
                initial.throwTest1Notify();
                test(false);
            } catch (notify ex) {
                // Expected
            }
```
`throwTest1Notify` throws a `notify` exception, and then we catch it.

After this we have tests that seem to be testing slicing in this context as well, with these 2 other operations:
```
    void throwTest1E2AsE1();
    void throwTest1E2AsE2();
```
But these seem like more advanced cases, whereas `notify` is testing a 'base case'.

File: java/test/src/main/java/test/Ice/adapterDeactivation/Test.ice (hunk 6 above)
Comment (InsertCreativityHere) [reply java/test/src/main/java/test/Ice/adapterDeactivation/Test.ice]: Fine, I changed it to `transientOp` in the places where it was absolutely necessary.
Is this our new escaping convention then? We suffix the illegal identifier with it's type?

ie. ` struct while { ... }` is mapped to `whileStruct`?

**Comment by InsertCreativityHere** [comment]
In the course of adding doc-comments, parameter names, and cleaning up code,
this PR also performs significant cleanup of `slice2java` in addition to implementing `java:identifier`.

File: java/test/src/main/java/test/Slice/escape/Key.ice (hunk 11 above)
Comment (InsertCreativityHere) [reply java/test/src/main/java/test/Slice/escape/Key.ice:11]: I updated all the escaped named in Java, C++, and C# to use this new convention.
Our preferred escaping prefix is `escaped` now.

Obviously I'm only talking about places where we can't use a built-in escaping thing like `@` in C#.
